drivers: wifi: siwx91x: Add AP configuration enhancements and hidden SSID support

ArunmaniAlagarsamy2710 opened the pull request:
This PR introduces enhancements to the siwx91x Wi-Fi driver, including support for configuring AP settings, hidden SSID functionality, and validation checks for interface state.

ArunmaniAlagarsamy2710 committed:
drivers: wifi: siwx91x: Add interface state validation

This patch introduces validation checks to ensure Wi-Fi commands are
executed only when the device is in a valid operational mode.

- Restricts command execution if the device is not in an appropriate mode
- Prevents reconfiguring the device when it is already in an active state
- Enabled Advanced multiprobe setting as default.

Signed-off-by: Arunmani Alagarsamy <arunmani.a@silabs.com>

NiteshKShah commented on drivers/wifi/siwx91x/siwx91x_wifi.c:
```diff
@@ -117,25 +145,67 @@ static int siwx91x_ap_enable(const struct device *dev, struct wifi_connect_req_p
+	}
+
 	if (params->band != WIFI_FREQ_BAND_UNKNOWN && params->band != WIFI_FREQ_BAND_2_4_GHZ) {
+		LOG_ERR("Invalid band");
```
Not supported

NiteshKShah commented on drivers/wifi/siwx91x/siwx91x_wifi.c:
```diff
@@ -117,25 +145,67 @@ static int siwx91x_ap_enable(const struct device *dev, struct wifi_connect_req_p
 	} else {
 		siwx91x_ap_cfg.channel.channel = params->channel;
 	}
 
```
Add check for not supporting bandwidth > 20 MHz

NiteshKShah commented on drivers/wifi/siwx91x/siwx91x_wifi.c:
```diff
@@ -117,25 +145,67 @@ static int siwx91x_ap_enable(const struct device *dev, struct wifi_connect_req_p
+
+	if (sidev->state == WIFI_STATE_COMPLETED) {
+		LOG_ERR("AP mode already enabled");
+		return -EINVAL;
```
EALREADY

NiteshKShah commented on drivers/wifi/siwx91x/siwx91x_wifi.c:
```diff
@@ -117,25 +145,67 @@ static int siwx91x_ap_enable(const struct device *dev, struct wifi_connect_req_p
+	ret = get_saved_ap_configuration(&saved_ap_cfg);
+	if (ret != SL_STATUS_OK) {
+		LOG_ERR("Failed to get the AP configuration: 0x%x", ret);
+		return -EINVAL;
```
Provide proper error returns

ArunmaniAlagarsamy2710 commented:
Rebased to resolve the conflicts

jerome-pouiller commented on drivers/wifi/siwx91x/siwx91x_wifi.c:
```diff
@@ -120,21 +106,41 @@ static int siwx91x_ap_enable(const struct device *dev, struct wifi_connect_req_p
 		return -ENOTSUP;
 	}
 
+	/*FIXME siwx91x_ap_cfg.channel.channel = SL_WIFI_BAND_2_4GHZ */
```
I don't understand this `FIXME`. `SL_WIFI_BAND_2_4GHZ` does not make sense for `siwx91x_ap_cfg.channel.channel`. Do you mean `siwx91x_ap_cfg.channel.band` ?

jerome-pouiller commented on drivers/wifi/siwx91x/siwx91x_wifi.c:
```diff
@@ -120,21 +106,41 @@ static int siwx91x_ap_enable(const struct device *dev, struct wifi_connect_req_p
+	}
+
+	if (sidev->state == WIFI_STATE_COMPLETED) {
+		return 0;
```
Do you know what the API expect if `ap_enable()` is called twice with different parameters?

ArunmaniAlagarsamy2710 replied:
Removed this change for now. We'll raise a separate PR to handle connection with different parameters

jerome-pouiller commented on drivers/wifi/siwx91x/siwx91x_wifi.c:
```diff
@@ -120,21 +106,41 @@ static int siwx91x_ap_enable(const struct device *dev, struct wifi_connect_req_p
+
+	siwx91x_ap_cfg.channel.bandwidth = SL_WIFI_BANDWIDTH_20MHz;
+
+	if ((params->ssid_length == 0) || (params->ssid_length > WIFI_SSID_MAX_LEN)) {
```
Parentheses are useless in this expression.

jerome-pouiller reviewed (commented):
Just a couple of questions. Otherwise, it looks fine.

jerome-pouiller commented on soc/silabs/silabs_siwx91x/siwg917/nwp.h:
```diff
@@ -9,17 +9,22 @@
+	uint8_t oper_mode;
+	bool hidden_ssid;
+	uint8_t max_num_sta;
+} siwx91x_boot_config_feature_t;
```
Unless you plan to add more parameters, I am not sure we need to introduce this struct. You can pass the 3 parameters to `siwx91x_nwp_mode_switch()`.

ArunmaniAlagarsamy2710 commented on drivers/wifi/siwx91x/siwx91x_wifi.c:
```diff
@@ -245,8 +259,13 @@ static int siwx91x_connect(const struct device *dev, struct wifi_connect_req_par
 	int ret;
 
+	if (sidev->state == WIFI_STATE_COMPLETED) {
+		return 0;
```
Removed this change for now. We'll raise a separate PR to handle connection with different parameters

jerome-pouiller commented on drivers/wifi/siwx91x/siwx91x_wifi.c:
```diff
@@ -303,7 +314,8 @@ static int siwx91x_connect(const struct device *dev, struct wifi_connect_req_par
 	memcpy(wifi_config.ssid.value, params->ssid, params->ssid_length);
 
-	ret = sl_wifi_connect(SL_WIFI_CLIENT_INTERFACE, &wifi_config, 0);
+	interface = sl_wifi_get_default_interface();
```
Sometime interface is initialized declaration, sometime later in the code. Is it possible to uniformize that?

jerome-pouiller commented on soc/silabs/silabs_siwx91x/siwg917/nwp.c:
```diff
@@ -20,9 +20,12 @@
 
-int siwg91x_get_nwp_config(int wifi_oper_mode, sl_wifi_device_configuration_t *get_config)
+int siwx91x_get_nwp_config(uint8_t wifi_oper_mode, bool hidden_ssid, uint8_t max_num_sta,
+			   sl_wifi_device_configuration_t *get_config)
```
Usually, the returned arguments are placed first in the list of arguments. Rationale, it mimics the usual code:
```
  memcpy(destination, source, ..);
  destination = source;
```

ArunmaniAlagarsamy2710 committed:
drivers: wifi: siwx91x: Replace opermode flags

Replaced SL_SI91X_CLIENT_MODE and SL_SI91X_ACCESS_POINT_MODE with
WIFI_STA_MODE and WIFI_SOFTAP_MODE, respectively, for AP configuration
command intergration.

Signed-off-by: Arunmani Alagarsamy <arunmani.a@silabs.com>

ArunmaniAlagarsamy2710 committed:
drivers: wifi: siwx91x: Enable AP configuration support

- Adds support for configuring client maximum inactivity timeout.
- Adds support for bandwidth, It supports 20MHZ only.
- Adds support for setting the maximum number of clients and
  hidden SSID mode by rebooting the NWP device.

Signed-off-by: Arunmani Alagarsamy <arunmani.a@silabs.com>

ArunmaniAlagarsamy2710 commented:
Rebased to resolve conflicts

ArunmaniAlagarsamy2710 committed:
drivers: wifi: siwx91x: Add Wi-Fi mgmt events

- Implement event handling for AP and STA modes
- Enable configurations for security (PSK), aggregation,
  and hidden PSK credentials

Signed-off-by: Arunmani Alagarsamy <arunmani.a@silabs.com>

ArunmaniAlagarsamy2710 commented:
Rebased to resolve conflicts

ArunmaniAlagarsamy2710 commented:
Introduced Wi-Fi management event callbacks to handle AP/STA events.

jerome-pouiller commented on drivers/wifi/siwx91x/siwx91x_wifi.c:
```diff
@@ -138,6 +180,7 @@ static int siwx91x_ap_enable(const struct device *dev, struct wifi_connect_req_p
 	struct siwx91x_dev *sidev = dev->data;
 	/* Wiseconnect requires a valid PSK even if WIFI_SECURITY_TYPE_NONE is selected */
 	static const char dummy_psk[] = "dummy_value";
+	sl_wifi_ap_configuration_t saved_ap_cfg = { };
```
This variable is going to be overwritten by `get_saved_ap_configuration()`, so it is not required to initialize it.

jerome-pouiller commented on drivers/wifi/siwx91x/siwx91x_wifi.c:
```diff
@@ -952,7 +1047,9 @@ static void siwx91x_iface_init(struct net_if *iface)
 
+	sidev->max_num_sta = CONFIG_WIFI_MGMT_AP_MAX_NUM_STA;
 	sidev->state = WIFI_STATE_INTERFACE_DISABLED;
+	sidev->hidden_ssid = false;
```
Already initialized.

jerome-pouiller commented on drivers/wifi/siwx91x/siwx91x_wifi.c:
```diff
@@ -952,7 +1047,9 @@ static void siwx91x_iface_init(struct net_if *iface)
 	struct siwx91x_dev *sidev = iface->if_dev->dev->data;
 	sl_status_t status;
 
+	sidev->max_num_sta = CONFIG_WIFI_MGMT_AP_MAX_NUM_STA;
```
You can probably initialize this field during the declaration of the variable.

ArunmaniAlagarsamy2710 replied:
The sidev structure is retrieved from `iface->if_dev->dev->data`, so it's already allocated before this function. Therefore, the `max_num_sta` field can't be initialized during declaration

jerome-pouiller commented on drivers/wifi/siwx91x/siwx91x_wifi.c:
```diff
@@ -811,6 +867,45 @@ static int siwx91x_mode(const struct device *dev, struct wifi_mode_info *mode)
+		if (params->bandwidth != WIFI_FREQ_BANDWIDTH_20MHZ) {
+			return -ENOTSUP;
+		}
+	}
```
Place this condition earlier with the other early exits.

jerome-pouiller commented on drivers/wifi/siwx91x/siwx91x_wifi.c:
```diff
@@ -28,6 +28,48 @@ LOG_MODULE_REGISTER(siwx91x_wifi);
 
 NET_BUF_POOL_FIXED_DEFINE(siwx91x_tx_pool, 1, _NET_ETH_MAX_FRAME_SIZE, 0, NULL);
 
+static int evaluate_nwp_reboot(const struct device *dev, uint8_t oper_mode)
```
You have prefix this function by `siwx91x`. Then this function does not only `evaluate`, it does the effective reboot. You may rename it in `siwx91x_nwp_reboot_if_required()` or something like this.

jerome-pouiller commented on soc/silabs/silabs_siwx91x/siwg917/nwp.c:
```diff
@@ -20,14 +20,22 @@
+	if (wifi_oper_mode == WIFI_SOFTAP_MODE && max_num_sta > AP_MAX_NUM_STA) {
+		LOG_ERR("Exceeded maximum supported stations (%d)", AP_MAX_NUM_STA);
+		return -EINVAL;
+	}
```
No statement before variable declarations please.

jerome-pouiller replied:
I expect to catch the error (with an assert?) if `(hidden_ssid == true || max_num_sta != 0) && wifi_oper_mode != WIFI_SOFTAP_MODE`.

jerome-pouiller commented on drivers/wifi/siwx91x/siwx91x_wifi.c:
```diff
@@ -70,6 +70,103 @@ static int siwx91x_set_max_sta(const struct device *dev, uint8_t max_sta)
+	default:
+		return "UNKNOWN";
+	}
+}
```
Replace that function with a table.

jerome-pouiller commented on soc/silabs/silabs_siwx91x/siwg917/nwp.c:
```diff
@@ -41,9 +41,12 @@ int siwx91x_get_nwp_config(sl_wifi_device_configuration_t *get_config, uint8_t w
 			.tcp_ip_feature_bit_map = SL_SI91X_TCP_IP_FEAT_EXTENSION_VALID,
-			.custom_feature_bit_map = SL_SI91X_CUSTOM_FEAT_EXTENSION_VALID,
+			.custom_feature_bit_map = (SL_SI91X_CUSTOM_FEAT_EXTENSION_VALID |
+						   SL_SI91X_CUSTOM_FEAT_ASYNC_CONNECTION_STATUS),
```
remove parentheses.

jerome-pouiller reviewed (commented):
The code start to become messy. The review is harder and harder. I am afraid we are going to reach a point where nothing could be changed anymore.

jerome-pouiller commented on drivers/wifi/siwx91x/siwx91x_wifi.c:
```diff
@@ -909,6 +964,40 @@ static int siwx91x_mode(const struct device *dev, struct wifi_mode_info *mode)
+		if (params->bandwidth != WIFI_FREQ_BANDWIDTH_20MHZ) {
+			return -ENOTSUP;
+		}
+	}
```
Place this check first and then call `get_saved_ap_configuration()`.

jerome-pouiller commented on soc/silabs/silabs_siwx91x/siwg917/nwp.c:
```diff
@@ -36,9 +36,12 @@ int siwx91x_get_nwp_config(sl_wifi_device_configuration_t *get_config, uint8_t w
 		.boot_option = LOAD_NWP_FW,
 		.boot_config = {
-			.feature_bit_map = SL_SI91X_FEAT_SECURITY_OPEN | SL_SI91X_FEAT_WPS_DISABLE,
+			.feature_bit_map = SL_SI91X_FEAT_SECURITY_PSK | SL_SI91X_FEAT_WPS_DISABLE |
```
Are we still able to connect to an open network without `SL_SI91X_FEAT_SECURITY_OPEN`?

jerome-pouiller commented on drivers/wifi/siwx91x/siwx91x_wifi.c:
```diff
@@ -70,6 +70,102 @@ static int siwx91x_set_max_sta(const struct device *dev, uint8_t max_sta)
+		}
+	}
+
+	return reason_strings[WLAN_REASON_UNKNOWN].str;
```
This does not work. I assume you can return `NULL` instead.

ArunmaniAlagarsamy2710 replied:
Instead of returning NULL, would it be acceptable to return the string literal "Unknown" directly? This way, we can avoid NULL checks at the call site.

jerome-pouiller replied:
yes, sure.

ArunmaniAlagarsamy2710 committed:
soc: silabs: siwg917: Refactor siwx91x_get_nwp_config

Refactor siwx91x_get_nwp_config function to reduce its
Cognitive Complexity from 34 to below the allowed limit of 25
as per SonarQube guidelines. Extracted configuration logic into
smaller helper functions.

Signed-off-by: Arunmani Alagarsamy <arunmani.a@silabs.com>

ArunmaniAlagarsamy2710 commented:
Resolved the sonarcube issues

jerome-pouiller commented on soc/silabs/silabs_siwx91x/siwg917/nwp.c:
```diff
@@ -109,66 +85,114 @@ int siwx91x_get_nwp_config(sl_wifi_device_configuration_t *get_config, uint8_t w
+		return -EINVAL;
 	}
+
+#ifdef CONFIG_WIFI_SILABS_SIWX91X
```
`if (IS_ENABLE(...)) ...`

jhedberg commented on soc/silabs/silabs_siwx91x/siwg917/nwp.c:
```diff
@@ -20,74 +20,63 @@
+	size_t sram_size = DT_REG_SIZE(DT_CHOSEN(zephyr_sram));
 
-	if (DT_REG_SIZE(DT_CHOSEN(zephyr_sram)) == KB(196)) {
+	if (sram_size == KB(196)) {
 		boot_config->ext_custom_feature_bit_map |= SL_SI91X_EXT_FEAT_480K_M4SS_192K;
-	} else if (DT_REG_SIZE(DT_CHOSEN(zephyr_sram)) == KB(256)) {
+	} else if (sram_size == KB(256)) {
 		boot_config->ext_custom_feature_bit_map |= SL_SI91X_EXT_FEAT_416K_M4SS_256K;
-	} else if (DT_REG_SIZE(DT_CHOSEN(zephyr_sram)) == KB(320)) {
+	} else if (sram_size == KB(320)) {
```
Actually, that PR modifies these same lines, so this will get sorted out when the merge conflict is fixed (depending on which PR goes in first).

ArunmaniAlagarsamy2710 replied:
Yes, it needs to be updated.
